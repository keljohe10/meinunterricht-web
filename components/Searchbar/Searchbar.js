import React, { useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import debounce from 'lodash.debounce';
import {
  Box,
  Input,
  InputGroup,
  InputLeftAddon,
  InputRightAddon,
  Flex,
} from '@chakra-ui/react';
import { SearchIcon, CloseIcon } from '@chakra-ui/icons';

import { getMovies } from '@/stores/movies';

const Searchbar = ({ typetext }) => {
  const dispatch = useDispatch();
  const [inputValue, setInputValue] = useState('');

  const debouncedFetch = useMemo(
    () => debounce(text => dispatch(getMovies(text)), 400),
    [dispatch],
  );

  const handleOnEnterKeyPressed = e => {
    if (e.key === 'Enter' && inputValue.length) debouncedFetch(inputValue);
  };

  const handleOnChange = value => {
    setInputValue(value);
    if (!value?.length || value.length >= 3) {
      debouncedFetch(value);
      typetext(true);
    }
  };

  const onClear = () => {
    setInputValue('');
    debouncedFetch();
  };

  return (
    <Flex flexDirection='column' h='full' bg='gray.900'>
      <Box w='full' position='relative' zIndex={10}>
        <Box
          w='full'
          align='center'
          pt={4}
          pb={6}
          mx={{ md: 'auto' }}
          px={{ base: 6, md: 0, lg: 4 }}
        >
          <Box position='relative' w={{ base: '100%', md: '2xl' }}>
            <InputGroup borderColor='gray.200'>
              <InputLeftAddon bg='white' px={3}>
                <SearchIcon color='alpha.400' h={4} w={4} />
              </InputLeftAddon>
              <Input
                value={inputValue}
                onChange={e => handleOnChange(e.target.value)}
                onKeyDown={e => handleOnEnterKeyPressed(e)}
                color='gray.600'
                placeholder='Search...'
                bg='white'
                _focus={{ boxShadow: 'none' }}
                data-testid='search-input'
              />
              {inputValue?.length && (
                <InputRightAddon
                  data-testid='clear-icon'
                  onClick={onClear}
                  bg='white'
                  cursor='pointer'
                >
                  <CloseIcon color='gray.500' h={3} w={3} />
                </InputRightAddon>
              )}
            </InputGroup>
          </Box>
        </Box>
      </Box>
    </Flex>
  );
};

Searchbar.propTypes = {
  typetext: PropTypes.func.isRequired,
};

export default Searchbar;
