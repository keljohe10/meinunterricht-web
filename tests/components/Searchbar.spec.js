/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom/extend-expect';

import { act, fireEvent, render, cleanup, waitFor } from '@/tests/testUtils';
import mockAdapter from '@/tests/mockAdapter';

import Searchbar from '@/components/Searchbar/Searchbar';
import Home from '@/pages/index';

beforeEach(cleanup);

describe('Searchbar component', () => {
  let wrapper;

  const movies = {
    movies: [
      {
        title: 'Race to Space',
        director: 'Sean McNamara',
        plot: 'In the 1960s a young woman works at NASA as an animal trainer responsible for the chimpanzee who will go into space.',
        poster:
          'https://m.media-amazon.com/images/M/MV5BMTczNDUwMDA1NV5BMl5BanBnXkFtZTcwNTA5MTAwMQ@@._V1_SX300.jpg',
        imbdID: 'tt0167360',
      },
      {
        title: 'Gore from Outer Space',
        director: 'Hirohisa Sasaki',
        plot: 'A woman is sent to execution for the murder of her daughter, who she believes is kidnapped. When she tells her story to a nun, we learn her story through a series of encounters with a psychic, a dirty politician, and two FBI agents who love Jell-O and believe there are aliens are responsible for bewildering the woman.',
        poster:
          'https://m.media-amazon.com/images/M/MV5BNTk2Nzg3NDEtMjQ4NC00Nzk4LWE0YmYtZTdmODYzYjIxOTljXkEyXkFqcGdeQXVyMzU0NzkwMDg@._V1_SX300.jpg',
        imbdID: 'tt0321649',
      },
    ],
  };

  beforeEach(() => {
    wrapper = render(
      <Home>
        <Searchbar typetext={() => true} />
      </Home>,
    );
  });

  it('matches snapshot', () => {
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('renders an input', () => {
    expect(wrapper.getByTestId('search-input')).toBeVisible();
  });

  it('renders a clear icon when theres some text and clears the input when pressed', async () => {
    const sampleText = 'someText';

    await act(async () => {
      fireEvent.change(wrapper.getByTestId('search-input'), {
        target: { value: sampleText },
      });
    });

    expect(wrapper.getByTestId('search-input').value).toBe(sampleText);
    expect(wrapper.getByTestId('clear-icon')).toBeVisible();

    await act(async () => {
      fireEvent.click(wrapper.getByTestId('clear-icon'));
    });

    expect(wrapper.getByTestId('search-input').value).toBe('');
  });

  it('renders 2 list elements when there is two results', async () => {
    const sampleText = 'sch';

    await act(async () => {
      fireEvent.change(wrapper.getByTestId('search-input'), {
        target: { value: sampleText },
      });
    });

    mockAdapter.onGet(`/movies`).reply(200, movies);

    await waitFor(
      () => {
        expect(wrapper.getByTestId('input-list')).toBeVisible();
        expect(wrapper.getAllByTestId('input-list-item').length).toBe(2);
      },
      {
        timeout: 1500,
      },
    );
  });
});
