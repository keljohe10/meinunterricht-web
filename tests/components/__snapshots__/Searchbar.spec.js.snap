// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-m7wkyr"
  >
    <div
      class="css-lygqz2"
    >
      <div
        class="css-79wky"
      >
        MeinUnterricht
      </div>
    </div>
  </div>
  <div
    class="css-0"
  >
    <div
      class="css-1k59v4j"
    >
      <div
        class="css-w8sp2g"
      >
        <div
          align="center"
          class="css-14g2rg7"
        >
          <div
            class="css-52xlms"
          >
            <div
              class="chakra-input__group css-8ejbcz"
            >
              <div
                class="chakra-input__left-addon css-1pw3sjv"
              >
                <svg
                  class="chakra-icon css-v6fbl0"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <input
                class="chakra-input css-1cjrfq4"
                data-testid="search-input"
                placeholder="Search..."
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1elhx01"
    >
      <div
        class="css-hoe9xz"
      >
        <div
          class="chakra-alert css-1adt752"
          role="alert"
        >
          <span
            class="chakra-alert__icon css-n0gcl4"
          >
            <svg
              class="chakra-icon css-fi0ww0"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
                fill="currentColor"
              />
            </svg>
          </span>
          Your search did not have any matches
        </div>
      </div>
    </div>
  </div>
  <span
    class="chakra-env"
    hidden=""
  />
</DocumentFragment>
`;
